src/compiler/core.ts

```diff
@@ -1394,14 +1394,28 @@ namespace ts {
         }
     }
 
-    export function copyListRemovingItem<T>(item: T, list: T[]) {
-        const copiedList: T[] = [];
-        for (const e of list) {
-            if (e !== item) {
-                copiedList.push(e);
+    /** Remove an item from an array, moving everything to its right one space left. */
+    export function removeItemAt<T>(array: T[], index: number): void {
+        // This seems to be faster than either `array.splice(i, 1)` or `array.copyWithin(i, i+ 1)`.
+        for (let i = index; i < array.length - 1; i++) {
+            array[i] = array[i + 1];
+        }
+        array.pop();
+    }
+
+    /** Remove the *first* occurrence of `item` from the array. */
+    export function removeItem<T>(item: T, array: T[]): void {
+        removeFirstItemWhere(array, element => element === item);
+    }
+
+    /** Remove the *first* element satisfying `predicate`. */
+    export function removeFirstItemWhere<T>(array: T[], predicate: (element: T) => boolean): void {
+        for (let i = 0; i < array.length; i++) {
+            if (predicate(array[i])) {
+                removeItemAt(array, i);

```

Comment:
This behavior seems different from the original `copyListRemovingItem`, as the old one removes *all* identical items, not just the first one